### PR TITLE
Searching in BST

### PR DESCRIPTION
**Pull request by pshivesh8**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Searching in BST

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue #4629

### CODE REVIEW

**Comment by siddharth25pandey** [comment]
@pshivesh8 Resolve the merge conflicts

**Comment by pshivesh8** [comment]
@siddharth25pandey  please tell me what is the conflict in merging?

**Comment by siddharth25pandey** [comment]
> @siddharth25pandey please tell me what is the conflict in merging?

@pshivesh8 The Conflicting file is **README.md** . It had arised when Git wasn't able to locate the **branch** where to push the changes. You can easily see the conflict by introspecting the **README** file you have commited

**Comment by pshivesh8** [comment]
@siddharth25pandey , I have resolved the conflict in README.md.
Please review

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+binary search tree is basically this traversal in which at each step we will go either towards left or right 
+and hence in at each step we discard one of the sub-trees. If the tree is balanced, we call a tree balanced if 
+for all nodes the difference between the heights of left and right subtrees is not greater than one, 
+we will start with a search space of n nodes and when we will discard one of the sub-trees 
+we will discard n/2 nodes so our search space will be reduced to n/2 and then in the next step 
+we will reduce the search space to n/4 and we will go on reducing like this till we find the element or 
+till our search space is reduced to only one node.
+*/
+
+#include<iostream>
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Use **bits/stdc++**

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		root->right = Insert_into_bst(root->right,data);
+	}
+
+	return root;
+}
+node* Create_BST(){
+
+	// root node
+	int data;				
+	cin >> data;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp:56]: Display a message in print form as what are you taking input from the user, for better UI

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+
+
+
+int main(){
+
+	node* root = Create_BST();
+	int ele;
+	cin >> ele;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Display a message in print form as what are you taking input from the user, for better UI

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	
+}
+
+
+
+
+int main(){
+
+	node* root = Create_BST();
+	int ele;
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Write a good naming convention

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	cin >> ele;
+    cout << boolalpha << search(root, ele);
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Display a message in print form as what are you showing output to the user, for better UI

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	else{
+		search(root->right, key);
+	}
+
+	
+}
+
+
+
+
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Remove the blank lines

**Review by siddharth25pandey — CHANGES_REQUESTED** [review]
@pshivesh8 Do the suggested changes

**Comment by pshivesh8** [comment]
@siddharth25pandey  I have resolved all those issues.
Please review

File: C-Plus-Plus/ds/Searching_in_BST.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+/* 
+	Test Case :
+
+	Input : 1 2 3 4 5 6 -1
+	4
+
+	Outout : true
+
```
Comment (siddharth25pandey) [inline C-Plus-Plus/ds/Searching_in_BST.cpp]: Sample I/O is **not matching** with the execution of the Program

**Review by siddharth25pandey — CHANGES_REQUESTED** [review]

**Comment by pshivesh8** [comment]
@siddharth25pandey , Please review
The code is working perfectly fine,  and I have attached the reference too :
### Note:  **You have to enter -1, at last, to stop inserting new nodes.**

image

**Comment by pshivesh8** [comment]
@siddharth25pandey  , I have updated the test cases too, Please review

**Review by siddharth25pandey — APPROVED** [review]
@pshivesh8 LGTM 💯 Able to merge

**Comment by siddharth25pandey** [comment]
@all-contributors please add @pshivesh8 for his code

**Comment by pshivesh8** [comment]
Thanks, @siddharth25pandey for guiding me  🎉🎉

**Pull request merged**